db_test.go

```diff
@@ -1303,6 +1303,100 @@ func TestInitializeHeadTimestamp(t *testing.T) {
 	})
 }
 
+func TestNoEmptyBlocks(t *testing.T) {
+	db, close := openTestDB(t, nil)
+	defer close()
+	defer db.Close()
+	db.DisableCompactions()
+
+	// Test no blocks after compact with empty head.
+	testutil.Ok(t, db.compact())
+	testutil.Equals(t, 0, len(db.Blocks()))
+
+	// Test no blocks after deleting all samples from head.
+	blockRange := DefaultOptions.BlockRanges[0]
+	label := labels.FromStrings("foo", "bar")
+
+	app := db.Appender()
+	for i := int64(0); i < 6; i++ {
+		_, err := app.Add(label, i*blockRange, 0)
+		testutil.Ok(t, err)
+		_, err = app.Add(label, i*blockRange+1000, 0)
+		testutil.Ok(t, err)
+	}
+	testutil.Ok(t, app.Commit())
+
+	testutil.Ok(t, db.Delete(math.MinInt64, math.MaxInt64, labels.NewEqualMatcher("foo", "bar")))
+	uid, err := db.compactor.Write(db.dir, db.head, db.head.MinTime(), db.head.MaxTime(), nil)
+	testutil.Ok(t, err)
+	testutil.Equals(t, ulid.ULID{}, uid)
+	testutil.Ok(t, db.reload())
+	testutil.Ok(t, db.head.Truncate(db.head.MaxTime()))
+	// No blocks created.
+	testutil.Equals(t, 0, len(db.Blocks()))
+
+	app = db.Appender()
+	for i := int64(7); i < 25; i++ {
+		for j := int64(0); j < 10; j++ {
+			_, err := app.Add(label, i*blockRange+j, 0)
+			testutil.Ok(t, err)
+		}
+	}
+	testutil.Ok(t, app.Commit())
+
+	testutil.Ok(t, db.compact())
```

Comment:
is there any reason why we are not using 
`db.compactor.Write(db.dir, db.head, db.head.MinTime(), db.head.MaxTime(), nil)` everywhere? I would rather avoid using block ranges as it is a bit unclear why we need something like `_, err := app.Add(label, i*blockRange+j, 0)` to trigger a new block creation. Also depending on `db.head.MaxTime()-db.head.MinTime() <= db.opts.BlockRanges[0]/2*3` will make the tests fragile if this logic changes.

Reply:
If we compact head manually using Write, we cannot come across many edge cases which I came across while using `db.compact()` (example running into infinite loop, not truncating head properly, etc). I would prefer keeping it like this, as this is the one called.

And I think block range is the range of time for a block? So if the samples in the head span multiple blocks, then head would be compacted. If in future with any changes new blocks were not created, then we would know by the assert below.

Reply:
in that case we should refactor something so that we don't have to rely in internal implementation and just test the behaviour.

Reply:
or maybe test one level up in the api.

Reply:
There is a check after this function call to make sure that new block was created. Would that be sufficient?

Reply:
maybe to make it more explicit use what I ended up doing in
https://github.com/prometheus/tsdb/blob/bac9cbed2eb417e0637ba2bed59d070b37deed44/db_test.go#L1468